example/build_cmake_rdc/functor.cpp

```diff
@@ -0,0 +1,21 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Macros.hpp>
+
+KOKKOS_EXTERNAL_FUNCTION void count_even(const long i, long& lcount) {
```

Comment:
I thought that macro would only be required on the extern declaration and that the definition would be a regular `KOKKOS_FUNCTION`.  Did I misunderstand how we are supposed to use it?

Reply:
> When a function is declared with SYCL_EXTERNAL, that macro must be used on the first declaration of that function in the translation unit. Redeclarations of the function in the same translation unit may optionally use SYCL_EXTERNAL, but this is not required.
>
> When a function is declared with SYCL_EXTERNAL, that function must also be defined in some translation unit, where the function is declared with SYCL_EXTERNAL.

Reads to me that you should annotate both the declaration and the definition. If I omit the annotation on the declaration, I am getting
```
/home/darndt/kokkos/example/build_cmake_rdc/cmake_example.cpp:44:54: error: SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
   44 |       n, KOKKOS_LAMBDA(const long i, long& lcount) { count_even(i, lcount); },
      |                                                      ^
/home/darndt/kokkos/example/build_cmake_rdc/cmake_example.cpp:22:22: note: 'count_even' declared here
   22 | KOKKOS_FUNCTION void count_even(const long i, long& lcount);
      |                      ^
/home/darndt/kokkos/example/build_cmake_rdc/cmake_example.cpp:44:10: note: called by 'operator()'
   44 |       n, KOKKOS_LAMBDA(const long i, long& lcount) { count_even(i, lcount); },
      |          ^
/home/darndt/kokkos-install/include/Kokkos_Macros.hpp:128:23: note: expanded from macro 'KOKKOS_LAMBDA'
  128 | #define KOKKOS_LAMBDA [=]
      |                       ^
1 error generated.
```
and if I'm omitting it on the definition, I'm seeing
```
terminate called after throwing an instance of 'sycl::_V1::compile_program_error'
  what():  The program was built for 1 devices
Build program log for 'Intel(R) Data Center GPU Max 1550':
Module <0x2c598f0>:  Unresolved Symbol <_Z10count_evenlRl>
Module <0x2c598f0>:  Unresolved Symbol <_Z10count_evenlRl>
Module <0x2c598f0>:  Unresolved Symbol <_Z10count_evenlRl>
Module <0x2c598f0>:  Unresolved Symbol <_Z10count_evenlRl> -11 (PI_ERROR_BUILD_PROGRAM_FAILURE)
Aborted
``` 
so I really need it in both places.

Reply:
That's interesting https://www.intel.com/content/www/us/en/developer/articles/technical/use-the-sycl-external-macro.html does not have it on the definition

Reply:
Maybe the `extern` was missing?

Reply:
That doesn't make a difference for me.

Reply:
Please confirm that you are saying the example from Intel's website gives an error at runtime

Reply:
A simple example with
```C++
// file1.cpp
#include <sycl/sycl.hpp>

SYCL_EXTERNAL int f();

int main() {
  sycl::queue q;
  q.single_task([](){ f(); });
  q.wait();
  std::cout << "done" << std::endl;
}
```
and
```C++
// file2.cpp
#include <sycl/sycl.hpp>

/*SYCL_EXTERNAL*/ int f() {
  return 2;
}
```
gives
```
> icpx -fsycl -c file1.cpp && icpx -fsycl -c file2.cpp && icpx -fsycl file1.o file2.o && ./a.out 
terminate called after throwing an instance of 'sycl::_V1::compile_program_error'
  what():  The program was built for 1 devices
Build program log for 'Intel(R) Data Center GPU Max 1550':
Module <0x2d54ec0>:  Unresolved Symbol <_Z1fv>
Module <0x2d54ec0>:  Unresolved Symbol <_Z1fv> -11 (PI_ERROR_BUILD_PROGRAM_FAILURE)
Aborted
```
You can first declare the function as `SYCL_EXTERNAL` in both compilation units but omit it in the definition, though:
```C++
// file2.cpp
#include <sycl/sycl.hpp>

SYCL_EXTERNAL int f();

int f() {
  return 2;
}
```
works.